app/dependencies/utils.py

```diff
@@ -1,18 +1,122 @@
 import json
 import logging
 import os
+import time
 from datetime import datetime
 
-import openai
 import tiktoken
 from dateutil.relativedelta import relativedelta
 from requests import Session
 from requests.exceptions import ConnectionError, Timeout, TooManyRedirects
 
-from app.dependencies.redis_client import set_redis_data
+from app.dependencies.redis_client import (
+    get_hexdata_from_redis,
+    set_redis_data,
+)
 
 
-def load_hex():
+def num_tokens_from_string(string: str, encoding_name: str) -> int:
+    """Returns the number of tokens in a text string."""
+    encoding = tiktoken.get_encoding(encoding_name)
+    num_tokens = len(encoding.encode(string))
+    return num_tokens
+
+
+def store_currency_in_redis():
+    url = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest"
+
+    rng = [1, 5001, 10001]
+    count = 1
+    for i in rng:
+        parameters = {"start": i, "limit": 5000, "convert": "USD", "aux": "cmc_rank"}
+
+        headers = {
+            "Accepts": "application/json",
+            "X-CMC_PRO_API_KEY": os.getenv("COIN_MARKET_CAP_KEY"),
+        }
+
+        session = Session()
+        session.headers.update(headers)
+
+        try:
+            response = session.get(url, params=parameters)
+            json_data = json.loads(response.text)
+
+            if count == 1:
+                set_redis_data("first_hexdata", json_data)
+            elif count == 2:
+                set_redis_data("second_hexdata", json_data)
+            else:
+                set_redis_data("third_hexdata", json_data)
+
+            count += 1
+
+        except (ConnectionError, Timeout, TooManyRedirects) as e:
+            logging.info(e)
+
+
+def get_currency_names():
+    curreny_ids_list = []
+    curreny_list = []
+
+    cryptocurrencies_first = get_hexdata_from_redis("first_hexdata")
+    cryptocurrencies_second = get_hexdata_from_redis("second_hexdata")
+    cryptocurrencies_third = get_hexdata_from_redis("second_hexdata")
+
+    temp = ""
+    count = 0
+    for i in range(0, len(cryptocurrencies_first)):
```

Comment:
can you explain what's going on in here?

Reply:
Here I am making strings to use in the historical endpoint because to send request in historical endpoint. our currency names should be like this "HEX,BTC,ETH" like in a single string comma seperated not like this ["btc","eth"] so i can't send the list directly to historical api. Also the cryptocurrencies_first have 5000 currency name so getting chunks from it by 100.

Reply:
why is it named get_hexdata_from_redis?

why are we again doing this
set_redis_data("curreny_list", curreny_list)

didn't we just get the same data from redis?

Reply:
In this line "set_redis_data("curreny_list", curreny_list)". I am setting the only list dictionaries of this format. ["name":"hex","price":"0.02",id,symbol]. to get only the price and check the currency name and symbol.
IN get_hexdata_from_redis we are getting the data of 10000 tokens of everything in raw json format data.